pool/pool.py

```diff
@@ -662,66 +674,72 @@ async def update_farmer_later():
-                )
-            else:
-                desired_state = PoolState(
-                    POOL_PROTOCOL_VERSION,
-                    PoolSingletonState.FARMING_TO_POOL.value,
-                    self.default_target_puzzle_hash,
-                    farmer_rec.singleton_tip_state.owner_pubkey,
-                    self.pool_url,
-                    self.relative_lock_height,
-                )
             singleton_task = asyncio.create_task(
-                get_and_validate_singleton_state_inner(
+                get_singleton_state(
                     self.node_rpc_client,
                     launcher_id,
                     farmer_rec,
-                    peak_height,
+                    self.blockchain_state["peak"].height,
                     self.confirmation_security_threshold,
-                    desired_state,
                 )
             )
             self.follow_singleton_tasks[launcher_id] = singleton_task
             remove_after = True
 
-        optional_result: Optional[Tuple[CoinSolution, PoolState, bool, bool]] = await singleton_task
+        optional_result: Optional[Tuple[CoinSolution, PoolState]] = await singleton_task
         if remove_after and launcher_id in self.follow_singleton_tasks:
             await self.follow_singleton_tasks.pop(launcher_id)
 
         if optional_result is None:
             return None
 
-        singleton_tip, singleton_tip_state, updated, is_pool_member = optional_result
-        if updated and remove_after:
+        singleton_tip, singleton_tip_state = optional_result
+
+        # Validate state of the singleton
+        is_pool_member = True
+        if singleton_tip_state.target_puzzle_hash != self.default_target_puzzle_hash:
+            self.log.info(f"Wrong target puzzle hash: {singleton_tip_state.target_puzzle_hash}")
+            is_pool_member = False
+        elif singleton_tip_state.relative_lock_height != self.relative_lock_height:
+            self.log.info(f"Wrong relative lock height: {singleton_tip_state.relative_lock_height}")
+            is_pool_member = False
+        elif singleton_tip_state.version != POOL_PROTOCOL_VERSION:
+            self.log.info(f"Wrong version {singleton_tip_state.version}")
+            is_pool_member = False
+        elif singleton_tip_state.state == PoolSingletonState.SELF_POOLING.value:
+            self.log.info(f"Invalid singleton state {singleton_tip_state.state}")
+            is_pool_member = False
+        elif singleton_tip_state.state == PoolSingletonState.LEAVING_POOL.value:
+            coin_record: Optional[CoinRecord] = await self.node_rpc_client.get_coin_record_by_name(singleton_tip.coin)
+            assert coin_record is not None
+            if self.blockchain_state["peak"].height - coin_record.confirmed_block_index > self.relative_lock_height:
+                is_pool_member = False
+
+        self.log.info(f"Is {launcher_id} pool member: {is_pool_member}")
+
+        if farmer_rec is not None and (
```

Comment:
This is not the same behaviour as before. Before, we were only updating if the new state was buried deeply enough in the chain. This is because a reorg will cause issues.

Reply:
Actually from my understanding the logic there should still be the same. The confirmations check is still here https://github.com/Chia-Network/pool-reference/blob/9c4fde0a0bbdfeb7c04e4c78a2cee411caf4b789/pool/singleton.py#L85-L88 it only returns the latest if it's deep enough so this check should only then be true, no?

Reply:
Yes you're right!